qiime/core/type/signature.py

```diff
@@ -9,18 +9,74 @@
 import collections
 import inspect
 import itertools
+import copy
 
 import qiime.sdk
 from .grammar import TypeExpression
 from .primitive import is_primitive_type
 from .semantic import is_semantic_type
 from .visualization import Visualization
+from ..util import ImmutableBase
+
+
+class _NoValue:
+    def __repr__(self):
+        return "NOVALUE"
+
+
+class ParameterSpec(ImmutableBase):
```

Comment:
Didn't know about the "default" ability, but it also seems a little unintuitive. I'd rather leave this a class I think.